llvm/lib/Analysis/UniformityAnalysis.cpp

```diff
@@ -32,7 +32,14 @@ bool llvm::GenericUniformityAnalysisImpl<SSAContext>::markDefsDivergent(
 
 template <> void llvm::GenericUniformityAnalysisImpl<SSAContext>::initialize() {
   for (auto &I : instructions(F)) {
-    if (TTI->isSourceOfDivergence(&I))
+    bool NoDivergenceSource = false;
+    if (auto Call = dyn_cast<CallInst>(&I)) {
+      if (Function *Callee = Call->getCalledFunction()) {
+        if (Callee->hasFnAttribute(Attribute::NoDivergenceSource))
```

Comment:
I think it might make more sense to check the attributes on the call instruction instead of the callee function, but I have never quite understood the interaction between the two -- e.g. is there something generic that propagates attributes from the callee to the call? Or are you supposed to check both the call and the callee (if known)?

Reply:
Checking attributes on the call will automatically check them on both the call itself and the callee, if it is known.

Reply:
Great. Then I think this patch should check attributes on the call. It seems reasonable that a call could be "no divergence source" either with an unknown callee (e.g. an indirect call) or because the frontend knows something about that specific call site even though the callee could in general introduce divergence.